### PR TITLE
Use Redis as session store

### PR DESCRIPTION
**Pull request by kiootic**
ref #1010

### CODE REVIEW

File: pkg/core/redis/context.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	"context"
+
+	"github.com/gomodule/redigo/redis"
+)
+
+type contextKeyType struct{}
+
+var contextKey = contextKeyType{}
+
+type Context interface {
```
Comment (louischan-oursky) [inline pkg/core/redis/context.go]: Seems unused in this commit.

File: pkg/core/auth/session/redis/store.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+func (s *store) Create(sess *session.Session) (err error) {
+	json, err := json.Marshal(sess)
+	if err != nil {
+		return
+	}
+	conn := redis.GetConn(s.ctx)
+	key := sessionKey(s.appID, sess.ID)
+	_, err = goredis.String(conn.Do("SET", key, json, "NX"))
+	if err == goredis.ErrNil {
+		err = fmt.Errorf("cannot create session")
```
Comment (louischan-oursky) [inline pkg/core/auth/session/redis/store.go]: Make this error constant?

File: pkg/core/auth/session/provider_impl_test.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+				CreatedAt:            initialTime,
+				AccessedAt:           initialTime,
+				AccessToken:          "access-token",
+				AccessTokenCreatedAt: initialTime,
+			}
+
+			Convey("should be successful", func() {
+				err := provider.Invalidate("session-id")
+				So(err, ShouldBeNil)
+				So(store.Sessions, ShouldBeEmpty)
```
Comment (louischan-oursky) [inline pkg/core/auth/session/provider_impl_test.go]: Get once before invalidation and assert the session is there.
Get once after invalidation and assert the session is gone.

File: pkg/core/auth/testing/context.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
+package testing
+
+import (
+	"github.com/skygeario/skygear-server/pkg/core/auth"
+	authinfo "github.com/skygeario/skygear-server/pkg/core/auth/authinfo"
+	authtoken "github.com/skygeario/skygear-server/pkg/core/auth/authtoken"
+	model "github.com/skygeario/skygear-server/pkg/core/model"
+)
+
+type MockContext struct {
```
Comment (louischan-oursky) [inline pkg/core/auth/testing/context.go]: This new API looks nice!

File: docker-compose.yml [hunk 5]
```diff
@@ -13,6 +13,10 @@ services:
     - db_data:/var/lib/postgresql/data
     ports:
     - "5432:5432"
+  redis:
+    image: redis:5.0
+    ports:
+      - "6379:6379"
```
Comment (louischan-oursky) [inline docker-compose.yml]: Make the data of this redis persistent.

File: pkg/core/redis/context.go (hunk 1 above)
Comment (kiootic) [reply pkg/core/redis/context.go]: It's some old prototyping code, will remove it.

File: pkg/core/auth/session/provider_impl_test.go (hunk 3 above)
Comment (kiootic) [reply pkg/core/auth/session/provider_impl_test.go]: I don't think it is needed? The parent block setup one session in the fake store, and each block assert the expected number of sessions in fake store after the operation.

**Pull request merged**